### PR TITLE
Client-Server Implementation of reCAPTCHA with React and Django

### LINKED ISSUE #4680: Client-Server Implementation of reCAPTCHA with React and Django
#### Client-Server Implementation of reCAPTCHA with React and Node.js
reCAPTCHA helps keep malicious users away from a website/web app and only allows access to legitimate users. It does this by posing a test that is easy for humans to solve and tricky for 'bots' to figure out. It is mostly used on forms to ensure that the data being submitted actually comes from a human being and not a bot.

In this article, the reader will learn how to integrate reCAPTCHA into a website so as to ensure that the site is not accessed by bots. We will use React to create the front end and Django for the backend. The frontend will be used to reCAPTCHA input and the backend will be used to verify it. Once the input is verified, the user will be able to submit the form.

#### Pre-requisites
- Understanding of the JavaScript programming language.
- Python and Django basics.
- Basic knowledge of React and React hooks.
- Understanding of the Fetch API.

#### Key takeaways 
- How to implement reCAPTCHA into your website.
- How to use the backend to verify the reCAPTCHA input and take the appropriate action.
- How to ensure that your site is not accessed by bots.

#### Article quality
My article is going to be different in the following ways:
- It will provide clear instructions on how to integrate the backend and the frontend.
- Display messages to the user on successful/unsuccessful reCAPTCHA resolving.
- Give a detailed explanation of the code used.

### PR DESCRIPTION
**Pull request by KayveTech**
This fixes  #4680

### CODE REVIEW

**Comment by KayveTech** [comment]
This fixes #4680

File: content/articles/client-server-react-django-recaptcha/index.md [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+#### Acquiring our Site and Secret Keys
+To enable reCAPTCHA on your website, you first need to register your site on the reCAPTCHA website so that you be able to acquire the site and secret keys. Click [here](https://www.google.com/recaptcha/admin/create) to open the Google reCAPTCHA section for registering a new site. Fill in the required information as shown below:
+
+- In the **Label** text field, enter your site name.
+- In the **reCAPTCHA** section, choose reCAPTCHA v2. In the options list that appears, select option 1("I'm not a robot" Checkbox).
+- In the **Domains** section, enter the following two options: 127.0.0.1 and localhost.
+- Accept the reCAPTCHA's terms of service and click submit.
+- You will then be provided with a **site key**(to use in the frontend) and a **secret key**(to use in the backend).
+
```
Comment (jumamiller) [inline content/articles/client-server-react-django-recaptcha/index.md:30]: Could you attach a screenshot for this part?

File: content/articles/client-server-react-django-recaptcha/index.md [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+</div>
+  );
+}
+
+export default App;
+```
+
+**Code explanation**
+
+- In the above code, we have a contact form that contains the fields; name, email, and message. We keep track of the data entered in these fields using the useState() hook.
```
Comment (jumamiller) [inline content/articles/client-server-react-django-recaptcha/index.md:121]: Have all methods enclosed with  backticks (` `) i.e the `useState()` hook.Update where applicable on the entire article.

File: content/articles/client-server-react-django-recaptcha/index.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+- In the `div` with the className `cta`, we have our ReCaptcha component that takes two attributes, one is the site key that we get from the Google reCAPTCHA website. The other attribute is an `onChange` attribute that calls the `handleRecaptcha` function when the reCAPTCHA component is clicked on.
+- Below the reCAPTCHA component, we conditionally render the `Submit` button depending on the outcome of the reCAPTCHA. We only show it when the success value of the reCAPTCHA evaluates to true(when the user successfully solves the reCAPTCHA challenge). This helps to ensure that the form data can only be submitted after the reCAPTCHA challenge has been solved.
+- The `handleRecaptcha` function takes in a value that we send to the backend for server-side verification. You can console.log the value to see it. We use the Fetch API to make a POST request to the route `http://127.0.0.1:8000/recaptcha/` which we will, later on, create on our server. We then use the response we get from the server and pass it to the `setCaptchaResult` function which sets the state of the `captchaResult`. The response is either true or false. If it is true, we show the `Submit` button.
```
Comment (jumamiller) [inline content/articles/client-server-react-django-recaptcha/index.md:124]: Add the opening and closing brackets for method definitions, i.e `handleRecaptcha()`.Update where applicable!

**Review by jumamiller — CHANGES_REQUESTED** [review]
Great article, fix minor suggestions and then we're good to move to tech review.

**Comment by KayveTech** [comment]
> Plagiarism check done ✅ Why did you delete our template before submitting your PR? @KayveTech

@ahmadmardeni1, I didn't delete the template. It didn't appear when I was opening the PR.

**Comment by KayveTech** [comment]
@jumamiller, I'll take care of the requested changes soon...

**Comment by KayveTech** [comment]
@jumamiller, requested changes done 🚀

**Comment by jumamiller** [comment]
Upload new image for 'recaptchaSnap.PNG '. The current one has a wrong format!

File: content/articles/client-server-react-django-recaptcha/index.md [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+To enable reCAPTCHA on your website, you first need to register your site on the reCAPTCHA website so that you be able to acquire the site and secret keys. Click [here](https://www.google.com/recaptcha/admin/create) to open the Google reCAPTCHA section for registering a new site. Fill in the required information as shown below:
+
+- In the **Label** text field, enter your site name.
+- In the **reCAPTCHA** section, choose reCAPTCHA v2. In the options list that appears, select option 1("I'm not a robot" Checkbox).
+- In the **Domains** section, enter the following two options: 127.0.0.1 and localhost.
+- Accept the reCAPTCHA's terms of service and click submit.
+- You will then be provided with a **site key**(to use in the frontend) and a **secret key**(to use in the backend).
+
+![Recaptcha](recaptchaSnap.PNG)
+
```
Comment (jumamiller) [inline content/articles/client-server-react-django-recaptcha/index.md:32]: Fix the image path

**Comment by KayveTech** [comment]
@jumamiller, is the image path okay now?

**Comment by jumamiller** [comment]
> @jumamiller, is the image path okay now?

Upload new image, the current one has wrong extension. or does it show on your end?

**Comment by hectorkambow** [comment]
> > @jumamiller, is the image path okay now?
> 
> Upload new image, the current one has wrong extension. or does it show on your end?

as long as the image is either .png or .jpg it should be fine (the image path needs to match). 👍

**Review by jumamiller — APPROVED** [review]

**Comment by hectorkambow** [comment]
jumping on this one

**Comment by hectorkambow** [comment]
check

**Comment by hectorkambow** [comment]
Very nice work gents @KayveTech 

cc @jumamiller  

Very easy to follow along 🚀  

keep up the excellent work.

**Pull request merged**